# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/wasm/WasmGC.cpp
# Commit: e6dbc240ccbb
# Full Hash: e6dbc240ccbb05b033eefca6563d73477d42c34c
# Author: Julien Pages <jpages@mozilla.com>
# Date: 2025-02-22 09:14:45
# Regressor Bug: 1861078
# File Overlap Count: 1
# Description:
#   Bug 1861078 - wasm: Refactor stackmaps to use a hashmap instead of a vector. r=rhunt
#   
#   Wasm stackmaps were stored in a sorted vector and traced at runtime with multiple
#   binary searches. Using hashmaps should reduce the complexity of iterating over stackmaps.
#   
# ==============================================================================

diff -r 9a162cc369e3 -r e6dbc240ccbb js/src/wasm/WasmGC.cpp
--- a/js/src/wasm/WasmGC.cpp	Fri Feb 21 17:17:58 2025 +0000
+++ b/js/src/wasm/WasmGC.cpp	Fri Feb 21 17:17:59 2025 +0000
@@ -315,3 +315,14 @@
 #  endif
 }
 #endif
+
+void StackMaps::checkInvariants(const uint8_t* base) const {
+#ifdef DEBUG
+    // Chech that each entry points from the stackmap structure points
+    // to a plausible instruction.
+    for (auto iter = mapping_.iter(); !iter.done(); iter.next()) {
+      MOZ_ASSERT(IsPlausibleStackMapKey(base + iter.get().key()),
+                "wasm stackmap does not reference a valid insn");
+    }
+#endif
+  }